File: Euler/part3/p133.py
#!/usr/bin/env python
#-*- coding:utf-8 -*-

from eulertools import primes

def f(n):
    """
    >>> f(100)
    454387123
    """
    return sum(p for p in primes(n) if not is_factor_R_10_n(p))


def is_factor_R_10_n(p):
    """测试p是否是R(10**n) {n 为自然数}的因子
    >>> is_factor_R_10_n(3)
    False
    >>> is_factor_R_10_n(11)
    True
    >>> is_factor_R_10_n(17)
    True
    >>> is_factor_R_10_n(43)
    True
    """
    # g(k) = sum(a[k] ** i for i in range(10)) % p
    # a[0] = 10
    # a[n] = 10 ** (10 ** n)
    # a[n + 1] = 10 ** (10 ** (n + 1)) = 10 ** ((10 ** n) * 10) = a[n] ** 10
    a = 10
    while True:
        result = sum(a ** i for i in range(10)) % p
        if result == 0:
            return True
        if a % p == 10 % p:
            return False
        a = a ** 10 % p
    # a[2] = 10 ** 100

File: Euler-Go/src/question/p133.go
package question
import "eulertools"

func exp_mod(a int64, b int64, p int64)int64{
    if b == 0{
        return 1
    }else if b % 2 == 0{
        tmp := exp_mod(a, b /2, p)
        return (tmp * tmp) % p
    }else{
        return (exp_mod(a, b - 1, p) * a) % p
    }
    return 0
}

func is_factor_of_R_n(p int64)bool{
    var a int64
    a = 10
    is_cycle := make(map[int64] bool)
    for{
        var sum int64
        sum =0
        for i :=0; i<10; i++{
            sum += exp_mod(int64(a), int64(i), p)
        }
        if(sum % p == 0){
            return true
        }
        _, in_cycle := is_cycle[a % p]
        if in_cycle{
            return false
        }
        is_cycle[a % p] = true
        a = exp_mod(a, 10, p)
    }
    return false
}

func P133(n int64)int64{
    //P133(100000)
    var result int64
    result = 0
    for _, p := range eulertools.Primes(n){
        if !is_factor_of_R_n(p){
            result += p
        }
    }
    return result
}
